codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java

```diff
@@ -190,6 +192,148 @@ public static void writeFrameHeader(ByteBuf out, int payloadLength, byte type,
     }
 
     /**
+     * Provides the ability to associate the outcome of multiple {@link ChannelPromise}
+     * objects into a single {@link ChannelPromise} object. This can be done with relatively low overhead
+     * if the number of {@link ChannelPromise} objects is known a priori.
+     */
+    static class SimpleChannelPromiseAggregator implements ChannelFutureListener {
+        private final ChannelHandlerContext ctx;
+        private final ChannelPromise promise;
+        private final int expectedCount;
+        private int completedCount;
+
+        SimpleChannelPromiseAggregator(ChannelHandlerContext ctx, int expectedCount, ChannelPromise promise) {
+            this.ctx = checkNotNull(ctx, "ctx");
+            this.promise = checkNotNull(promise, "promise");
+            if (expectedCount < 0) {
+                throw new IllegalArgumentException("expectedCount < 0");
+            }
+            this.expectedCount = expectedCount;
+            if (expectedCount == 0) {
+                promise.setSuccess();
+            }
+        }
+
+        public ChannelPromise newPromise() {
+            return ctx.newPromise().addListener(this);
+        }
+
+        public ChannelPromise tryFailure(Throwable cause) {
+            // Prevent future notifications from marking this promise as failed or successful
+            completedCount = expectedCount;
+            promise.tryFailure(cause);
+            return promise;
+        }
+
+        @Override
+        public void operationComplete(ChannelFuture future) throws Exception {
+            if (completedCount < expectedCount) {
+                ++completedCount;
+                if (!future.isSuccess()) {
+                    tryFailure(future.cause());
+                    return;
+                }
+                if (completedCount == expectedCount) {
+                    promise.setSuccess();
+                }
+            }
+        }
+    }
+
+    static ChannelFuture tryFailure(Throwable cause, SimpleChannelPromiseAggregator promiseAggregator,
+            ChannelPromise promise) {
+        if (promiseAggregator == null) {
+            return promise.setFailure(cause);
+        } else {
+            return promiseAggregator.tryFailure(cause);
+        }
+    }
+
+    static ChannelFuture writeUnsignedShort(ChannelHandlerContext ctx, int value, ChannelPromise promise) {
```

Comment:
Maybe you can just change `writeFrameHeader` (below) to create one buffer for the header since it's only a few bytes?  Do we know if there would be any performance advantage to one approach vs the other?